internal/crypto/src/cose/certificate_acceptance_policy.rs

```diff
@@ -0,0 +1,260 @@
+    /// >
+    /// > * ~~The list of X.509 certificate trust anchors provided by the C2PA
+    /// > (i.e., the C2PA Trust List).~~ _(not relevant for this API)_
+    /// > * ~~A list of additional X.509 certificate trust anchors.~~ _(not
+    /// > relevant for this API)_
+    /// > * A list of accepted Extended Key Usage (EKU) values.
+    /// >
+    /// > NOTE: Some of these lists can be empty.
+    ///
+    /// This function reads zero or more EKU object identifiers (OIDs) in
+    /// dotted-decimal notation (one per line) and configures the trust handler
+    /// to accept certificates that are issued with one of those EKUs.
+    ///
+    /// IMPORTANT: The trust configuration will always accept the default set of
+    /// OIDs descfibed in the C2PA Technical Specification.
+    ///
+    /// This function will quietly ignore any invalid input, such as a non-UTF8
+    /// input or lines within the input such as comments or blank lines that can
+    /// not be parsed as OIDs.
+    ///
+    /// [§14.4.1, C2PA Signers]: https://c2pa.org/specifications/specifications/2.1/specs/C2PA_Specification.html#_c2pa_signers
+    pub fn add_valid_ekus(&mut self, eku_oids: &[u8]) {
+        let Ok(eku_oids) = std::str::from_utf8(eku_oids) else {
+            return;
+        };
+
+        for line in eku_oids.lines() {
+            if let Ok(_oid) = Oid::from_str(line) {
+                self.additional_ekus.insert(line.to_string());
+            }
+        }
+    }
+
+    /// Remove all trust anchors, private credentials, and EKUs previously
+    /// configured.
+    pub fn clear(&mut self) {
+        self.trust_anchor_ders.clear();
+        self.end_entity_cert_ders.clear();
+        self.additional_ekus.clear();
+    }
+
+    /// Return an iterator over the trust anchors.
+    ///
+    /// Each anchor will be returned in DER format.
+    pub fn trust_anchor_ders(&self) -> impl Iterator<Item = &'_ Vec<u8>> {
+        self.trust_anchor_ders.iter()
+    }
+
+    /// Return an iterator over the allowed end-entity certificates.
+    ///
+    /// Each end-entity certificate will be returned in DER format.
+    pub fn end_entity_cert_ders(&self) -> impl Iterator<Item = &'_ Vec<u8>> {
+        self.end_entity_cert_ders.iter()
+    }
+
+    /// Return `true` if the EKU OID is allowed.
+    pub fn has_allowed_eku<'a>(&self, eku: &'a ExtendedKeyUsage) -> Option<Oid<'a>> {
+        if eku.email_protection {
+            return Some(EMAIL_PROTECTION_OID.clone());
+        }
+
+        if eku.time_stamping {
+            return Some(TIMESTAMPING_OID.clone());
+        }
+
+        if eku.ocsp_signing {
+            return Some(OCSP_SIGNING_OID.clone());
+        }
+
+        // TO REVIEW: Earlier implementation used the last match; this one uses the
+        // first. Does that make a difference?
+
+        for extra_oid in eku.other.iter().as_ref() {
+            let extra_oid_str = extra_oid.to_string();
```

Comment:
I don't think this behavior is correct.  Each "." in the pattern means something.  Matching a subset is incorrect.  Look at the old code.

Reply:
`self.additional_ekus` is defined as `HashSet<String>`. Unless I've misinterpreted that data type, `contains` should only return `true` if the **entire** string presented is an exact match for one of the strings in the hash set. I don't think it's prone to the subset match that you're concerned about.

Reply:
The one thing that is known to be different from the old code is that this will return the _first_ OID in the cert's OID list that matches rather than the last. I couldn't find a case where that was important, but if it is, I can adapt.